pkg/kubelet/kuberuntime/utils.go

```diff
@@ -0,0 +1,184 @@
+func parsePodSandboxName(name string) (podName, podNamespace, podUID string, err error) {
+	parts := strings.Split(name, "_")
+	if len(parts) == 0 || parts[0] != podSandboxNamePrefix {
+		err = fmt.Errorf("failed to parse pod sandbox name %q into parts", name)
+		return "", "", "", err
+	}
+	if len(parts) < 5 {
+		glog.Warningf("found a pod sandbox with the %q prefix, but too few fields (%d): %q", podSandboxNamePrefix, len(parts), name)
+		err = fmt.Errorf("Pod sandbox name %q has less parts than expected %v", name, parts)
+		return "", "", "", err
+	}
+
+	return parts[1], parts[2], parts[3], nil
+}
+
+// Creates a name which can be reversed to identify both full pod name and container name.
+// This function returns stable name, unique name and an unique id.
+func buildContainerName(podName, podNamespace string, container *api.Container) (string, string, string) {
+	containerName := container.Name + "." + strconv.FormatUint(kubecontainer.HashContainer(container), 16)
+	stableName := fmt.Sprintf("%s_%s_%s_%s",
+		containerNamePrefix,
+		containerName,
+		podName,
+		podNamespace)
+	UID := fmt.Sprintf("%08x", rand.Uint32())
+	return stableName, fmt.Sprintf("%s_%s", stableName, UID), UID
+}
+
+// Unpacks a container name, returning the pod full name and container name
+func parseContainerName(name string) (podName, podNamespace, containerName string, hash uint64, err error) {
+	// Some container runtimes appear to be appending '/' to names.
+	name = strings.TrimPrefix(name, "/")
+	parts := strings.Split(name, "_")
+	if len(parts) == 0 || parts[0] != containerNamePrefix {
+		err = fmt.Errorf("failed to parse container name %q into parts", name)
+		return "", "", "", 0, err
+	}
+	if len(parts) < 5 {
+		glog.Warningf("found a container with the %q prefix, but too few fields (%d): %q", containerNamePrefix, len(parts), name)
+		err = fmt.Errorf("Container name %q has less parts than expected %v", name, parts)
+		return "", "", "", 0, err
+	}
+
+	nameParts := strings.Split(parts[1], ".")
+	containerName = nameParts[0]
+	if len(nameParts) > 1 {
+		hash, err = strconv.ParseUint(nameParts[1], 16, 32)
+		if err != nil {
+			glog.Warningf("invalid container hash %q in container %q", nameParts[1], name)
+		}
+	}
+
+	return parts[2], parts[3], containerName, hash, nil
+}
+
+func milliCPUToShares(milliCPU int64) int64 {
+	if milliCPU == 0 {
+		// Docker converts zero milliCPU to unset, which maps to kernel default
+		// for unset: 1024. Return 2 here to really match kernel default for
+		// zero milliCPU.
+		return minShares
+	}
+	// Conceptually (milliCPU / milliCPUToCPU) * sharesPerCPU, but factored to improve rounding.
+	shares := (milliCPU * sharesPerCPU) / milliCPUToCPU
+	if shares < minShares {
+		return minShares
+	}
+	return shares
+}
+
+func toRuntimeProtocol(protocol api.Protocol) runtimeApi.Protocol {
+	switch protocol {
+	case api.ProtocolTCP:
+		return runtimeApi.Protocol_TCP
+	case api.ProtocolUDP:
+		return runtimeApi.Protocol_UDP
+	}
+
```

Comment:
Print something here or else we don't even know the input is unexpected.
In the original code, we printed:
```
glog.Warningf("Unknown protocol %q: defaulting to TCP", port.Protocol)
```